### ISSUE TITLE
OAuth Authorization Limits 

### ISSUE THREAD

**Issue by ernee** [google/site-kit-wp#1115]
Prompted by [this support topic](https://wordpress.org/support/topic/25-site-authorization-limit/), this issue is a continuation of internal discussions on possible authorization limits and the proxy implementation of Site Kit that needs further confirmation.

**Comment by adamsilverstein** [comment]
@ernee, I looked into this a little bit. 

In the original thread the author points to (https://wordpress.org/support/topic/25-active-tokens-site-limit/) the response explains that the access limitation here is account based, eg. that API access is limited by user account. It looks the like the refresh token limit (50) is still true: https://developers.google.com/identity/protocols/OAuth2#expiration and would apply to any way a user access google APIs.

Although I haven't tried to reproduce this issue, I assume these types of access issues would affect Site Kit as well. _This is an API limitation, not directly a Site Kit limitation._ 

For example if the user set up 51 websites with Site Kit, all using the same google username, the first site would like have an invalidated refresh token. Once the main token expired (max 1 hr), they would need to re-authorize to continue using Site Kit in wp-admin on that site.

Since I haven't actually reproduced this issue I would say its fair to call this an expected limitation of Site Kit.

**Recommendation:**
It sounds like this one person is setting up Analytics for many different websites, I'm guessing the common use case of a webmaster or agency setting up sites for clients. 

In terms of accessing Analytics data, probably best practice would be to _grant the site owner's Google account access to the Analytics Property_ in Analytics itself. Then they can access the connection through their account directly once they log into WordPress. If each site owner connects using their own account, the sites per user limitation shouldn't prevent the work of getting Site Kit set up.

If it turns out this is a common use case, we might want consider documenting it somewhere directly eg. "How to set up Site Kit for someone else on their website".

**Comment by ernee** [comment]
thanks for the info @adamsilverstein ! Is the 25 refresh token _per service_ account limit mentioned also at play? I'm assuming [this is the reference](https://developers.google.com/analytics/devguides/config/mgmt/v3/authorization#invalid_grant) for Analytics and [a similar one](https://developers.google.com/tag-manager/api/v2/authorization#invalid-grant) for Tag Manger. 

> For example if the user set up 51 websites with Site Kit, all using the same google username, the first site would like have an invalidated refresh token. Once the main token expired (max 1 hr), they would need to re-authorize to continue using Site Kit in wp-admin on that site.

Just to confirm: is there a set user experience for those that need to re-authenticate? Although rare, I've seen notices in the dashboard asking for "re-authentication" or even perhaps "issue accessing data". Would be great to know what to expect at this point (cc: @felixarntz).

**Comment by adamsilverstein** [comment]
I'm still a little unclear how these limitations come into play. Do we have confirmations from users that this is an issue or only that they suspect it is an issue? I'd like to try to reproduce, however I'm unclear exactly what will cause the limitation to come into play.

**Comment by adamsilverstein** [comment]
I was able to reproduce this issue with Site Kit as well. I set up Site Kit on around 28 sites using the same Google account over a period of a couple of days and eventually lost access to data on the earlier sites, requiring that I re-login to regain access. As you anticipated, that same API limits affect your account access via Site Kit, since these are account/API limits.

I left a comment on the original issue - https://wordpress.org/support/topic/25-site-authorization-limit/#post-12456822

I suggest closing this issue as "Won't Fix" - essentially this is an Account/API limitation, not a Site Kit limitation. The one thing I think we could do is **create a dedicated guide/section in our docs explaining how "webmasters" should set up multiple/many sites for their clients** and properly share access with them (eg. via the google accounts themselves currently.)

**Comment by adamsilverstein** [comment]
@ernee We recently resolved a bug that revoked tokens on all previous sites when the user hit the limit. The original limit still applies though, users will only be able to remain connected to 25 site/services at a time. 

It is important to note that Site Kit will still place the required/specified tags on the user's site (so they are still collecting analytics data), the authorization is only required for the user to view data in Site Kit.

I'm going to leave this open until I have a chance to repeat my 26 site test, then close this as resolved and comment back on the original support ticket.

**Comment by adamsilverstein** [comment]
I re-tested this by setting up 27 sites, adding analytics all using the same account. Everything worked well this time, I was no longer kicked out of **any** of the sites. I didn't try going to 50 or more sites where another potential threshold exists (the setup process is time consuming).  

I also tested disconnecting from the user menu of one site and verifying this no longer logged me out of other sites.

Closing this as resolved, this issue should be fixed for the vast majority of users. If users do get disconnected (their refresh token revoked) due to the user/account limits, they will no longer get logged out across their sites. 

I will update the support thread.
